private/buf/cmd/buf/buf.go

```diff
@@ -317,14 +317,47 @@ func wrapError(err error) error {
 		connectCode := connectErr.Code()
 		switch {
 		case connectCode == connect.CodeUnauthenticated, isEmptyUnknownError(err):
-			if authErr, ok := bufconnect.AsAuthError(err); ok && authErr.TokenEnvKey() != "" {
+			loginCommand := "buf registry login"
+			authErr, ok := bufconnect.AsAuthError(err)
+			if !ok {
+				// This code should be unreachable.
+				return fmt.Errorf("Failure: you are not authenticated. "+
+					"Set the %[1]s environment varialbe or run %q, using a Buf API Key as the password. "+
+					"If you have set the %[1]s or already have an entry in your netrc, "+
+					"check to see that your token is not expired. "+
+					"For details, visit https://buf.build/docs/bsr/authentication",
+					bufconnect.TokenEnvKey,
+					loginCommand,
+				)
+			}
+			// Invalid token found in env var.
+			if authErr.HasToken() && authErr.TokenEnvKey() != "" {
 				return fmt.Errorf("Failure: the %[1]s environment variable is set, but is not valid. "+
-					"Set %[1]s to a valid Buf API key, or unset it. For details, "+
-					"visit https://docs.buf.build/bsr/authentication", authErr.TokenEnvKey())
+					"Set %[1]s to a valid Buf API key, or unset it. "+
+					"For details, visit https://buf.build/docs/bsr/authentication",
+					authErr.TokenEnvKey(),
+				)
+			}
+			if authErr.Remote() != bufconnect.DefaultRemote {
+				loginCommand = fmt.Sprintf("%s %s", loginCommand, authErr.Remote())
 			}
-			return errors.New("Failure: you are not authenticated. Create a new entry in your netrc, " +
-				"using a Buf API Key as the password. If you already have an entry in your netrc, check " +
-				"to see that your token is not expired. For details, visit https://docs.buf.build/bsr/authentication")
+			// Invalid token found in netrc.
+			if authErr.HasToken() {
+				return fmt.Errorf("Failure: invalid token for %s found in netrc. "+
```

Comment:
Since we are not suggesting manually editing `netrc`, should it be mentioned here anymore? It could become `Failure: your API token is invalid. ...`

Reply:
yea i would change to that

Reply:
resolved in [321d525](https://github.com/bufbuild/buf/pull/3093/commits/321d52528e8885353230059ab93dcd66e1c8fdbe)